File: aoc-2015/src/main/java/com/capital7software/aoc/aoc2015aoc/days/Day20.java
package com.capital7software.aoc.aoc2015aoc.days;

import com.capital7software.aoc.lib.AdventOfCodeSolution;
import com.capital7software.aoc.lib.math.InfiniteHouses;
import java.time.Instant;
import java.util.List;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * --- Day 20: Infinite Elves and Infinite Houses ---<br><br>
 * To keep the Elves busy, Santa has them deliver some presents by hand, door-to-door.<br>
 * He sends them down a street with infinite houses numbered sequentially: 1, 2, 3, 4, 5,
 * and so on.
 *
 * <p><br>
 * Each Elf is assigned a number, too, and delivers presents to houses based on that number:
 * <ul>
 *     <li>
 *         The first Elf (number 1) delivers presents to every house: 1, 2, 3, 4, 5, ....
 *     </li>
 *     <li>
 *         The second Elf (number 2) delivers presents to every second house: 2, 4, 6, 8, 10, ....
 *     </li>
 *     <li>
 *         Elf number 3 delivers presents to every third house: 3, 6, 9, 12, 15, ....
 *     </li>
 * </ul>
 * There are infinitely many Elves, numbered starting with 1.<br>
 * Each Elf delivers presents equal to ten times his or her number at each house.
 *
 * <p><br>
 * So, the first nine houses on the street end up like this:<br>
 *
 * <p><br>
 * <code>
 * House 1 got 10 presents.<br>
 * House 2 got 30 presents.<br>
 * House 3 got 40 presents.<br>
 * House 4 got 70 presents.<br>
 * House 5 got 60 presents.<br>
 * House 6 got 120 presents.<br>
 * House 7 got 80 presents.<br>
 * House 8 got 150 presents.<br>
 * House 9 got 130 presents.<br>
 * </code>
 * <ul>
 *     <li>
 *         The first house gets 10 presents: it is visited only by Elf 1, which delivers
 *         1 * 10 = 10 presents.
 *     </li>
 *     <li>
 *         The fourth house gets 70 presents, because it is visited by Elves 1, 2, and 4,
 *         for a total of 10 + 20 + 40 = 70 presents.
 *     </li>
 * </ul>
 * What is the lowest house number of the house to get at least as many presents as the
 * number in your puzzle input?
 *
 * <p><br>
 * Your puzzle answer was 776160.
 *
 * <p><br>
 * --- Part Two ---<br><br>
 * The Elves decide they don't want to visit an infinite number of houses. Instead,
 * each Elf will stop after delivering presents to 50 houses. To make up for it,
 * they decide to deliver presents equal to eleven times their number at each house.
 *
 * <p><br>
 * With these changes, what is the new lowest house number of the house to get
 * at least as many presents as the number in your puzzle input?
 *
 * <p><br>
 * Your puzzle answer was 786240.
 */
public class Day20 implements AdventOfCodeSolution {
  private static final Logger LOGGER = LoggerFactory.getLogger(Day20.class);

  /**
   * Instantiates the solution instance.
   */
  public Day20() {

  }

  @Override
  public String getDefaultInputFilename() {
    return "inputs/input_day_20-01.txt";
  }

  @Override
  public void runPart1(List<String> input) {
    var start = Instant.now();
    int deliveredPresents = Integer.parseInt(input.getFirst());
    var lowest = lowestHouseNumber(deliveredPresents);
    var end = Instant.now();
    LOGGER.info("The lowest house number to be delivered {} presents is: {}",
                deliveredPresents, lowest);
    logTimings(LOGGER, start, end);
  }

  @Override
  public void runPart2(List<String> input) {
    var start = Instant.now();
    int deliveredPresents = Integer.parseInt(input.getFirst());
    var lowest = lowestHouseNumberNewRules(deliveredPresents);
    var end = Instant.now();
    LOGGER.info("Using the new rules, the lowest house number to be delivered {} presents is: {}",
                deliveredPresents, lowest);
    logTimings(LOGGER, start, end);
  }

  /**
   * Returns the lowest house number based on the number of delivered presents.
   *
   * @param deliveredPresents The number of delivered presents.
   * @return The lowest house number based on the number of delivered presents.
   */
  public int lowestHouseNumber(int deliveredPresents) {
    var infiniteHouses = new InfiniteHouses(deliveredPresents);

    return infiniteHouses.lowestHouseNumber();
  }

  /**
   * Returns the lowest house number based on the number of delivered presents
   * using the new rules.
   *
   * @param deliveredPresents The number of delivered presents.
   * @return The lowest house number based on the number of delivered presents
   *     using the new rules.
   */
  public int lowestHouseNumberNewRules(int deliveredPresents) {
    var infiniteHouses = new InfiniteHouses(deliveredPresents);

    return infiniteHouses.lowestHouseNumberNewRules();
  }
}

File: aoc-2017/src/main/kotlin/com/capital7software/aoc/aoc2017aoc/days/Day20.kt
package com.capital7software.aoc.aoc2017aoc.days

import com.capital7software.aoc.lib.AdventOfCodeSolution
import com.capital7software.aoc.lib.geometry.Particle3D
import com.capital7software.aoc.lib.geometry.ParticleSwarm
import com.capital7software.aoc.lib.geometry.Point3D
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import java.time.Instant
import org.slf4j.Logger
import org.slf4j.LoggerFactory

/**
 * **--- Day 20: Particle Swarm ---**
 *
 * Suddenly, the GPU contacts you, asking for help. Someone has asked it to simulate
 * **too many particles**, and it won't be able to finish them all in time to render
 * the next frame at this rate.
 *
 * It transmits to you a buffer (your puzzle input) listing each particle in order
 * (starting with particle 0, then particle 1, particle 2, and so on). For each particle,
 * it provides the X, Y, and Z coordinates for the particle's position (p), velocity (v),
 * and acceleration (a), each in the format <X,Y,Z>.
 *
 * Each tick, all particles are updated simultaneously. A particle's properties are
 * updated in the following order:
 *
 * - Increase the X velocity by the X acceleration.
 * - Increase the Y velocity by the Y acceleration.
 * - Increase the Z velocity by the Z acceleration.
 * - Increase the X position by the X velocity.
 * - Increase the Y position by the Y velocity.
 * - Increase the Z position by the Z velocity.
 *
 * Because of seemingly tenuous rationale involving z-buffering, the GPU would like to know
 * which particle will stay closest to position <0,0,0> in the long term. Measure this using
 * the Manhattan distance, which in this situation is simply the sum of the absolute values
 * of a particle's X, Y, and Z position.
 *
 * For example, suppose you are only given two particles, both of which stay entirely on
 * the X-axis (for simplicity). Drawing the current states of particles 0 and 1 (in that order)
 * with an adjacent a number line and diagram of current X positions (marked in parentheses),
 * the following would take place:
 *
 * ```
 * p=< 3,0,0>, v=< 2,0,0>, a=<-1,0,0>    -4 -3 -2 -1  0  1  2  3  4
 * p=< 4,0,0>, v=< 0,0,0>, a=<-2,0,0>                         (0)(1)
 *
 * p=< 4,0,0>, v=< 1,0,0>, a=<-1,0,0>    -4 -3 -2 -1  0  1  2  3  4
 * p=< 2,0,0>, v=<-2,0,0>, a=<-2,0,0>                      (1)   (0)
 *
 * p=< 4,0,0>, v=< 0,0,0>, a=<-1,0,0>    -4 -3 -2 -1  0  1  2  3  4
 * p=<-2,0,0>, v=<-4,0,0>, a=<-2,0,0>          (1)               (0)
 *
 * p=< 3,0,0>, v=<-1,0,0>, a=<-1,0,0>    -4 -3 -2 -1  0  1  2  3  4
 * p=<-8,0,0>, v=<-6,0,0>, a=<-2,0,0>                         (0)
 * ```
 *
 * At this point, particle 1 will never be closer to <0,0,0> than particle 0, and so,
 * in the long run, particle 0 will stay closest.
 *
 * **Which particle will stay closest to position <0,0,0>** in the long term?
 *
 * Your puzzle answer was **376**.
 *
 * **--- Part Two ---**
 *
 * To simplify the problem further, the GPU would like to remove any particles that **collide**.
 * Particles collide if their positions ever **exactly match**. Because particles are updated
 * simultaneously, **more than two particles** can collide at the same time and place.
 * Once particles collide, they are removed and cannot collide with anything else after that tick.
 *
 * For example:
 *
 * ```
 * p=<-6,0,0>, v=< 3,0,0>, a=< 0,0,0>
 * p=<-4,0,0>, v=< 2,0,0>, a=< 0,0,0>    -6 -5 -4 -3 -2 -1  0  1  2  3
 * p=<-2,0,0>, v=< 1,0,0>, a=< 0,0,0>    (0)   (1)   (2)            (3)
 * p=< 3,0,0>, v=<-1,0,0>, a=< 0,0,0>
 *
 * p=<-3,0,0>, v=< 3,0,0>, a=< 0,0,0>
 * p=<-2,0,0>, v=< 2,0,0>, a=< 0,0,0>    -6 -5 -4 -3 -2 -1  0  1  2  3
 * p=<-1,0,0>, v=< 1,0,0>, a=< 0,0,0>             (0)(1)(2)      (3)
 * p=< 2,0,0>, v=<-1,0,0>, a=< 0,0,0>
 *
 * p=< 0,0,0>, v=< 3,0,0>, a=< 0,0,0>
 * p=< 0,0,0>, v=< 2,0,0>, a=< 0,0,0>    -6 -5 -4 -3 -2 -1  0  1  2  3
 * p=< 0,0,0>, v=< 1,0,0>, a=< 0,0,0>                       X (3)
 * p=< 1,0,0>, v=<-1,0,0>, a=< 0,0,0>
 *
 * ------destroyed by collision------
 * ------destroyed by collision------    -6 -5 -4 -3 -2 -1  0  1  2  3
 * ------destroyed by collision------                      (3)
 * p=< 0,0,0>, v=<-1,0,0>, a=< 0,0,0>
 * ```
 *
 * In this example, particles 0, 1, and 2 are simultaneously destroyed at the time and
 * place marked X. On the next tick, particle 3 passes through unharmed.
 *
 * **How many particles are left** after all collisions are resolved?
 *
 * Your puzzle answer was **574**.
 */
class Day20 : AdventOfCodeSolution {
  private companion object {
    private val log: Logger = LoggerFactory.getLogger(Day20::class.java)
  }

  override fun getDefaultInputFilename(): String = "inputs/input_day_20-01.txt"

  override fun runPart1(input: MutableList<String>) {
    val start = Instant.now()
    val answer = getParticleClosestToTargetFast(input)
    val end = Instant.now()

    log.info("$answer is particle that will stay closest to the target in the long run!")
    logTimings(log, start, end)
  }

  override fun runPart2(input: MutableList<String>) {
    val start = Instant.now()
    val answer = getSurvivingParticles(input)
    val end = Instant.now()

    log.info("${answer.size} is the number of surviving particles!")
    logTimings(log, start, end)
  }

  /**
   * Returns the ID of the particle that will stay closest to the target in the long run.
   *
   * @param input The particles to parse.
   * @return The ID of the particle that will stay closest to the target in the long run.
   */
  @SuppressFBWarnings
  fun getParticleClosestToTarget(input: List<String>): Int {
    val targetPoint = Point3D(0L, 0L, 0L)
    val target: Particle3D<Long> = Particle3D(targetPoint, targetPoint, targetPoint)
    val instance = ParticleSwarm(input, target)
    return instance.closestInTheLongRun().id
  }

  /**
   * Returns the ID of the particle that will stay closest to the target in the long run.
   *
   * @param input The particles to parse.
   * @return The ID of the particle that will stay closest to the target in the long run.
   */
  fun getParticleClosestToTargetFast(input: List<String>): Int {
    val targetPoint = Point3D(0L, 0L, 0L)
    val target: Particle3D<Long> = Particle3D(targetPoint, targetPoint, targetPoint)
    val instance = ParticleSwarm(input, target)
    return instance.closestInTheLongRunFast().id
  }

  /**
   * Returns the ID of the particle that will stay closest to the target in the long run.
   *
   * @param input The particles to parse.
   * @return The ID of the particle that will stay closest to the target in the long run.
   */
  fun getSurvivingParticles(input: List<String>): List<ParticleSwarm.SwarmParticle> {
    val targetPoint = Point3D(0L, 0L, 0L)
    val target: Particle3D<Long> = Particle3D(targetPoint, targetPoint, targetPoint)
    val instance = ParticleSwarm(input, target)
    return instance.particlesAfterCollisions()
  }
}
